### PR TITLE
NERO overhaul

### PR DESCRIPTION
**Pull request by gjacobrobertson**
This is a big Pull Request which encompasses a large number of changes and significantly refactors many part of the codebase. The overarching goal of these changes is to enable experimenters to more easily design new experiments and AI algorithms, primarily in the NERO environment. I will be attempting to document the many proposed changes here in subsequent comments.

### CODE REVIEW

**Comment by gjacobrobertson** [comment]
Design Goals
==========

Loose Coupling
-----------------------------

Python module and class responsibilities should be clear, contained, and only depend on the interfaces between components, not their implementations
  * Implementing new Nero agent AIs should consist solely of updating `NERO/agent.py`
  * There should be a consistent interface for team level AIs (such as RTNEAT)
  * There should be a consistent method of serializing and deserializing NERO agents and teams that is flexible to the design of new Agent and Team classes
  * The NERO environment should completely define the environment without reference to the NERO module.
  * The NERO environment should be agnostic to agent implementations
  * The NERO module should define an interface for interacting with the environment
  * The C++ RTNEAT AI should not depend on environment implementations

Flexible Experimenting
------------------------------

Module and experiment designers should be able to use existing algorithms in a flexible way
  * Experiments making use of RTNEAT should have complete control over the fitness function being used.


NERO Learning Improvements
------------------------------------------
Existing NERO agents should converge towards player defined goals reasonably quickly, and the user interface should provide meaningful insight into progress being made
  * The NERO environment should report weighted reward instead of raw fitness values. Reward weights are functionally environmental state.
  * The NERO environment should report bounded rewards for potential use in algorithms that require bounded rewards (e.g. RL techniques with neural networks)
  * An agent's reported "time alive" hint should not exceed its maximum lifespan.
  * For any non-zero setting to reward weights, agent fitness should be diverse amongst the population over time, i.e. the majority of agents should not have 0 fitness the majority of the time.
  * Reward values should be consistent across reward weight configurations. i.e. changing reward weights should not change the scale or bounds of rewards given, which might invalidate prior or current settings.
  * Agents should be capable of learning and evolving in a team vs. team battle scenario

Code Quality
------------------
Code should be readable and maintainable
  * Commented out code snippets, unused functions, and unreachable sections of code should be removed. That's what version control is for.

**Comment by gjacobrobertson** [comment]
Changes
=======

Agent Lifecycle
--------------------

* Any agent AI is completely encapsulated by a subclass of `NeroAgent`
* `RTNEATAgent` has been renamed to `NEATAgent` which represents any agent that is backed by a C++ `NEAT::Organism`. It acts by feeding its sensors into the inputs of a neural network, activating the network, and interpreting the outputs as an action vector. It is agnostic to population level AI i.e. it can be used on its own, as part of RTNEAT, generational NEAT, or any other team.
* ` agent.py` provides factory methods for creating NeroAgent objects from string descriptions of AIs to be used. e.g. `agent.factory('neat') will return a new `NEATAgent`


Team Lifecycle
--------------------

* new `NeroTeam` class represents a set of agents in an environment and any population level AI for that team. It can be instantiated on its own to be a plain team of agents with no evolution or other population level methods.
* `teams.py` provides factory methods for creating Team objects
  * `create_agent(agent_ai, *args)` calls agent factory methods to instantiate a new agent and add it to the team
  * `create_agents(agent_ai)` fills the entire population with newly created agents 
  * `is_episode_over(agent)` hook into the environment lifecycle.
  * `reset(agent)` hook into the environment lifecycle
  * `kill_agent(agent)` marks an agent as dead
  * `reset_all()` resets an entire team to a playable state, resurrects dead agents, etc
  * `is_destroyed()` returns True iff the team is non-empty and all agents on the team have been killed
  * `start_training()` starts any training AI's
  * `stop_training()` stops any training AI's

* RTNEATTeam represents a `NEAT::Population` managed by the RTNEAT algorithm. It uses `is_episode_over` to hook agents killed periodically by the RTNEAT algorithm into the simulation lifecycle, and `reset(agent)` to replace organisms killed by RTNEAT. `reset_all` is used for generational NEAT to turn the population over the next epoch.

**Comment by gjacobrobertson** [comment]
TODO
=====

  * Update Roomba module to use new RTNEAT interface
  * Update Tabular Q-Learning Agent for new interfaces.
  * BUG: user's moving team spawn points no longer works.
  * `NEAT::Population::epoch` can segfault when all organisms' fitness is 0. This is currently avoided by not calling `epoch` in that situation, but the `epoch` function is 500+ lines of code and could probably use some cleaning anyhow
  * Update Battle Mode user interface for setting reward weights, saving teams, etc
  * MAYBE: It's possible that Training and Battle no longer even have to be distinct modules, but could share a single interface, and include something along the lines of checkboxes as to what kinds of AI's should be running, e.g. Team Training checkbox, Agent Learning checkboxes, evolution, etc.
  * BONUS: If we unify battle and training into a single interface maybe we could completely lose Java as a dependency

File: mods/NERO/agent.py [hunk 1]
```diff
@@ -1,202 +1,74 @@
...
-        """
-        return self.key_action()
-
-    def end(self, time, reward):
-        """
-        take in last reward
-        """
-        # store last reward and add average reward to fitness (i.e. how good the agent was since
-        # collecting the last cube)
+        self.rewards = init_info.sensors
```
Comment (ikarpov) [inline mods/NERO/agent.py]: it seems weird to me that you are assigning init_info.sensors to self.rewards... what does that mean?

File: mods/NERO/constants.py [hunk 2]
```diff
@@ -46,7 +46,9 @@
 pop_size = 50
 
 # number of steps per lifetime
-DEFAULT_LIFETIME = 1000
+DEFAULT_LIFETIME_MAX = 10000
+DEFAULT_LIFETIME_MIN = 1000
```
Comment (ikarpov) [inline mods/NERO/constants.py]: maybe comment what these mean here...

File: mods/NERO/agent.py (hunk 1 above)
Comment (gjacobrobertson) [inline mods/NERO/agent.py]: Looks like a bug. Thanks.

**Pull request merged**